docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+There are two ways of storing the discriminator: `wide packing` and `NaN-boxing`.
+
+### Wide packing
+
+This is the naive representation, using an 8-bit descriptor and the largest object's size as the payload.
+
+Example:
+```ts
+(A | U64) -> (Pointer | i64) -> { i8, i64 } // On both 32-bit and 64-bit machines
+// On CHERI, this would be { i8, i128 }, as pointers are 128-bits wide
+```
+
+This can be used for all types, but it creates two main problems:
+1. It adds at least 8-bits to every number.
+   For a `U8`, that's twice the size.
+2. Alignment requirements may force us to _pad_ the first `i8`.
+   This would add another 24, 56 or 120 bits on structures or arrays, per element.
+
+For this reason, we use NaN-boxing below for all types that we can.
+The encoding of the discriminator is shown below.
+
+### NaN-Boxing
+
+IEEE-754 doubles are 64 bits, on both 32-bit and 64-bit platforms, encoded in the exact same way:
+```
+[ sign ][ exponent ][ mantissa ]
+63     62          51          0
+```
+So:
+* 1 sign bit
+* 11 exponent bits
+* 52 mantissa bits
+
+A `NaN` (not-a-number) is encoded in IEEE-754 as setting the exponent to _all ones_.
+All other bits are ignored.
+If we set the exponent to _all ones_, we can use the remaining 53 bits to encode a variety of smaller types.
+
+Basically, if the exponent is **not** _all ones_, then the value represented is a 64-bit floating pointer number.
+If it is, then we use some bits to encode which type it is (the discriminator), and the rest of the bits left to encode the value.
+
+We use the sign bit to represent pointers, leaving the remaining 52-bits (the entire mantissa) to represent the address of the pointer.
+Most 64-bit architectures use 52-bits or less to encode addresses, so that's enough for our purposes.
+When the sign bit is zero, we use 4 high bits in the mantissa to encode all representable types, followed by a padding of `(52-4-sizeof(type))` bits, and the value taking the `sizeof(type)` lower bits.
+
+We can encode all 32-bit or less types as well as pointers and `F64`, but not 64-bit integers or higher.
+For those, we use the wide packing below.
+
+The list of the types that can be NaN-packed:
+* Pointer
+* F64
+* F32
+* U32
+* U16
+* U8
+* I32
+* I16
+* I8
+* Bool
+* ISize (on a 32 bit platform)
+* USize (on a 32 bit platform)
+
+In the following ways:
+```
+[*][***********][52 bit mantissa] = F64
+[1][11111111111][52 data bits] = Pointer
+[0][11111111111][0000][47*][1 data bit] = Bool
+[0][11111111111][0001][16*][32 data bits] = F32
+[0][11111111111][0010][40*][8 data bits] = I8
+[0][11111111111][0011][32*][16 data bits] = I16
+[0][11111111111][0100][16*][32 data bits] = I32
+[0][11111111111][0101][16*][32 data bits] = ISize (32)
+[0][11111111111][0110][40*][8 data bits] = U8
+[0][11111111111][0111][32*][16 data bits] = U16
+[0][11111111111][1000][16*][32 data bits] = U32
+[0][11111111111][1001][16*][32 data bits] = USize (32)
+```
```

Comment:
Does Verona's `F64` type exclude `NaN` values?  (If not, presumably we need at least one "`NaN`-boxed" `NaN` to serve as that value, and whenever we commit a `F64` to the heap we need to canonicalize any possible FPU-generated `NaN`s to that/one of those `NaN`s?)

Reply:
I thought about that but didn't discuss on meetings or wrote anything to that effect here.

We can have a 4bit-pattern for "actual NaNs", like `[1111]`.

Alternatively, actual `NaN`s can (almost) always be converted to all-zeroes on the mantissa before storing in a union.
* If the type is `F64`, it's still a `NaN`.
* If the type was a pointer, it's a NULL pointer and thus invalid in Verona.
* If the type is a `Bool`, then it could be a `false`, so we could distinguish `false` by setting some other bit to 1.
* Any other type won't allow all-zeroes as a representation.

But this adds a check on every floating point before storing in a union. Ideally, we'd be able to differentiate betweenn quiet and trapping `NaN`s if the hardware only emits one, we use the other for all other values. This may not work everywhere (or even anywhere), but it would remove the need to check every FP value, including most often non-`NaN` values.

Reply:
Actually, if we made `Bool`'s pattern `[0001]` and made sure no pattern is `[0000]`, it would just work without hacks.